File: Python/206. Reverse Linked List.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# noinspection PyMethodMayBeStatic, PyPep8Naming
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        return self.reverse(head, None)

    def reverse(self, head: Optional[ListNode], next: Optional[ListNode]) -> Optional[ListNode]:
        if head is None:
            return next

        current = head.next
        head.next = next

        return self.reverse(current, head)

File: JavaScript/206. Reverse Linked List.js
// noinspection JSUnusedGlobalSymbols
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
const reverseList = (head) => {
  return reverse(head, null)
}

const reverse = (head, next) => {
  if (head === null) {
    return next
  }

  const current = head.next
  head.next = next

  return reverse(current, head)
}
